lib/cn/kvset_builder.c

```diff
@@ -339,30 +348,40 @@ kvset_builder_finish(struct kvset_builder *imp)
 {
     merr_t err = 0;
 
-    assert(imp->kbb);
-    assert(imp->vbb);
+    INVARIANT(imp->hbb);
+    INVARIANT(imp->kbb);
+    INVARIANT(imp->vbb);
 
-    err = kbb_finish(imp->kbb, &imp->kblk_list, imp->seqno_min, imp->seqno_max);
+    err = kbb_finish(imp->kbb, &imp->kblk_list);
     if (ev(err))
         return err;
 
     if (imp->kblk_list.n_blks == 0) {
-        /* There are no kblocks. This happens when each input key has
-         * a tombstone and we are in "drop_tomb" mode.  This
-         * output kvset is empty and should not be created.  Destroy
-         * the corresponding vblock generator (which aborts any
-         * mblocks it has already allocated) and move on.  The empty
-         * kblk_list will prevent htis kvset from being created. */
+        /* There are no kblocks. This happens when each input key has a
+         * tombstone and we are in "drop_tomb" mode. This output kvset is empty
+         * and should not be created. Destroy the corresponding hblock vblock
+         * builder (which aborts any mblocks they had already allocated) and
+         * move on. The empty kblk_list will prevent this kvset from being
+         * created.
+         */
         vbb_destroy(imp->vbb);
-        imp->vbb = 0;
-        return 0;
+        imp->vbb = NULL;
+        goto out;
     }
 
     err = vbb_finish(imp->vbb, &imp->vblk_list);
     if (ev(err))
         return err;
 
-    return 0;
+out:
+    /* A constructed kvset will always only have 1 vgroup */
```

Comment:
Correct me if I am wrong.

Reply:
Before, if there were no kblocks, then there was no kvset.  That's no longer always tree because prefix tombs are stored in the hblock.  So now you can have a kvset with no kblocks.  I'd have to think a bit on how to handle that scenario here.

Important: the rest of our code must be checked to ensure it doesn't assume kvsets have kblocks.

Reply:
I don't know enough to say...

Reply:
Looks good. Easier code to follow. Implemented in next revision.